File: iOverlayAppRust/src/test/irregular_polygon_test.rs
use std::f64::consts::PI;
use std::time::Instant;
use i_overlay::bool::fill_rule::FillRule;
use i_overlay::bool::overlay_rule::OverlayRule;
use i_overlay::layout::overlay::{Overlay, ShapeType};
use crate::test::util::Util;

pub(crate) struct IrregularPolygonTest;

// 100_000 - 0.017672999
// 200_000 - 0.048966016
// 300_000 - 0.09378231
// 400_000 - 4.907101405

// Two irregular self-intersecting polygons are generated, the vertices of which are defined by a fixed radius and angle.
impl IrregularPolygonTest {
    pub(crate) fn run(n: usize, rule: OverlayRule) {
        println!("Start IrregularPolygon Test 2");
        let subj_path = Util::irregular_polygon(1000.0, 0.0, n);
        let clip_path = Util::irregular_polygon(1000.0, 0.5 * PI, n);

        let start = Instant::now();

        let mut overlay = Overlay::new(n);
        overlay.add_path(&subj_path, ShapeType::Subject);
        overlay.add_path(&clip_path, ShapeType::Clip);

        let graph = overlay.build_graph(FillRule::NonZero);
        let result = graph.extract_shapes(rule);

        assert!(!result.is_empty());

        let duration = start.elapsed();
        println!("Count: {:?}, time: {:?}", n, duration);
    }
}

File: ClipperApp/IrregularPolygonTest.cpp
//
// Created by Nail Sharipov on 31.01.2024.
//

#include "IrregularPolygonTest.h"
#include "Util.h"
#include "include/clipper2/clipper.h"
#include <iostream>
#include <chrono>
#include <cassert>

using namespace Clipper2Lib;

// Two irregular self-intersecting polygons are generated, the vertices of which are defined by a fixed radius and angle.
void IrregularPolygonTest::run(int n, ClipType clipType) {
    std::cout << "Start IrregularPolygon Test\n";

    auto poly0 = irregularPolygon(1000.0, 0.0, n);
    auto poly1 = irregularPolygon(1000.0, 0.5 * PI, n);

    Paths64 subj;
    subj.push_back(poly0);

    Paths64 clip;
    clip.push_back(poly1);

    auto start = std::chrono::high_resolution_clock::now();

    Paths64 solution = BooleanOp(clipType, FillRule::NonZero, subj, clip);

    assert(!solution.empty());

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;

    std::cout << "Count: " << n << ", time: " << elapsed.count() << "\n";
}